File: articles/analytical-chemistry-identification-matter/julia/analytical_kernel.jl
# Analytical chemistry numerical kernel in Julia.

function concentration_from_calibration(signal, slope, intercept)
    return (signal - intercept) / slope
end

function lod(blank_sd, slope)
    return 3.0 * blank_sd / slope
end

function loq(blank_sd, slope)
    return 10.0 * blank_sd / slope
end

function chromatographic_resolution(tR1, tR2, w1, w2)
    return 2.0 * (tR2 - tR1) / (w1 + w2)
end

function beer_lambert_concentration(absorbance, epsilon, path_length)
    return absorbance / (epsilon * path_length)
end

println("unknown_concentration=", round(concentration_from_calibration(3.72, 0.515, 0.04), digits=6))
println("LOD=", round(lod(0.0032, 0.515), digits=6))
println("LOQ=", round(loq(0.0032, 0.515), digits=6))
println("resolution=", round(chromatographic_resolution(3.10, 5.20, 0.42, 0.50), digits=6))
println("beer_lambert_c=", beer_lambert_concentration(0.625, 12500.0, 1.0))

File: articles/analytical-chemistry-identification-matter/fortran/analytical_kernel.f90
program analytical_kernel
  implicit none

  real(8) :: concentration, detection_limit, quantification_limit
  real(8) :: resolution, beer_c

  concentration = (3.72d0 - 0.04d0) / 0.515d0
  detection_limit = 3.0d0 * 0.0032d0 / 0.515d0
  quantification_limit = 10.0d0 * 0.0032d0 / 0.515d0
  resolution = 2.0d0 * (5.20d0 - 3.10d0) / (0.42d0 + 0.50d0)
  beer_c = 0.625d0 / (12500.0d0 * 1.0d0)

  print *, "unknown_concentration=", concentration
  print *, "LOD=", detection_limit
  print *, "LOQ=", quantification_limit
  print *, "resolution=", resolution
  print *, "beer_lambert_c=", beer_c

end program analytical_kernel
